modifying approach to ignoring app insights coming from workers

brettsam opened the pull request:
#8915 introduced a way to ignore logs coming from App Insights based on an app setting. This takes that approach and modifies it so that it will also work for a worker that declares this behavior via a capability at startup.

Rather than relying strictly on the app setting, this now looks at a ~custom property on the current `Activity`~. This value is set in the grpc channel and is based on that channel's capabilities (or on the app setting as before).

Edit: Approach changed to using an `AsyncLocal`.

jviau commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -49,7 +49,7 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
         private readonly WaitCallback _processInbound;
         private readonly object _syncLock = new object();
-        private readonly Dictionary<MsgType, Queue<PendingItem>> _pendingActions = new ();
+        private readonly Dictionary<MsgType, Queue<PendingItem>> _pendingActions = new Dictionary<MsgType, Queue<PendingItem>>();
```
This change necessary? I personally like the `new();` syntax (because the type is clear from the left side).

brettsam replied:
What settings do I need to build this locally? Something that I have configured automatically removes that space, then it fails to build, then I re-add it, etc. But it only happens if I ever open this file. I literally cannot run locally with this like this, and this was the easiest solution.

I suspect others would hit this as well... if there's some editorconfig I can use to make this work, I'd gladly do it.

I tried moving to a newer StyleCop, but that opened up a whole new can of worms with dozens of newer violations that it wanted me to address.

jviau replied:
That space _should_ be removed. The style is typically `new()` and not `new ()`. What error are you getting for the failed build? I see we don't have a `global.json` to lock the dotnet SDK in place, so this could be a case of you having an older dotnet SDK?

brettsam replied:
This was fixed in #8975. Have rebased on that change.

jviau commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -916,18 +930,38 @@ internal void Log(GrpcEvent msg)
-                            context.Logger.Log(logLevel, new EventId(0, rpcLog.EventId), rpcLog.Message, exception, (state, exc) => state);
+                            if (stateTuple.AppInsightsEnabledOnWorker)
+                            {
+                                if (Activity.Current != null)
```
I am not super familiar with our `Activity` usage, but I know `Activity` has optimizations when using `ActivitySource` to _not create_ the Activity at all when there are no listeners. So, do we risk `Activity` _sometimes_ being `null` here and thus this will work only some of the time?

I see that above we have a comment that "this allows AsyncLocal state to flow to loggers". What do you think of using our own `AsyncLocal<T>` here instead? Might not need the cleanup logic as well then. Or if you do, the `T` could be a small `IDisposable` implementation which performs the cleanup on dispose.

brettsam replied:
None of this uses ActivitySource today -- it's all manually created by App Insights and/or WebJobs.

Your comment about `AsyncLocal` is what I had originally proposed to @fabiocav (wrapped in something like `IHttpContextAccessor`... where another class can access some saved state of the current invocation). Is that something like what you were thinking?

I'm thinking about moving to that now after your comment. Any odd dependencies we take on Activity like this may be troublesome to decouple in the future. For example, if no one is listening, then stuff starts to fall apart b/c we don't have an Activity to store this state on.

jviau replied:
You could do some more robust accessor similar to `IHttpContextAccessor` if you want to put in that effort. I was thinking more a pointed `AsyncLocal` which is for _just_ this use case. It can be a static and directly accessed by both places that set and get it. So long as we keep it `internal` I am not worried too much about the robustness of it. It is just as you said - relying on `Activity` may not be the most stable approach. Especially if we make any changes to our tracing in the future.

jviau commented on src/WebJobs.Script/Config/ApplicationInsightsLoggerFilterOptionsSetup.cs:
```diff
@@ -0,0 +1,32 @@
+            // set this AsyncLocal to true to indicate the log should be skipped.
+            if (!_environment.IsDotNetInProc())
+            {
+                options.AddFilter<ApplicationInsightsLoggerProvider>((_, _) => !FilterApplicationInsightsFromWorker.Value);
```
Given the pains we have had with provider-specific filters, I do wonder if this will also prevent the non-provider filters from being applied?

Can we verify if when this is `True`, that other filters will still be applied?

brettsam replied:
Yes, I can add a test for this. To be specific, is your concern that b/c this filter is specific, it gets picked and other filters are ignored so if a worker does not have a capability, suddenly something generic like this in host.json would be ignored?

```
{
  "Logging": {
    "LogLevel": {
      "Function.Function1.User": "None" 
    }
  }
}
```

brettsam replied:
Okay -- good call here as it did completely interact with the other filters in non-intuitive ways. I've gone ahead and moved the check to an `ITelemetryProcessor` now, meaning that only App Insights will see this and all filters remain as-is.

mathewc commented on src/WebJobs.Script/ScriptConstants.cs:
```diff
@@ -39,6 +39,8 @@ public static class ScriptConstants
         public const string LogPropertyProcessIdKey = "ProcessId";
         public const string LogPropertyActivityIdKey = "MS_ActivityId";
 
+        public const string IgnoreApplicationInsightsKey = "MS_IgnoreApplicationInisghts";
```
Is this used anywhere? I don't see a reference.

brettsam replied:
Not any more; i've removed it in my next commit  -- also spelled wrong!

mathewc commented:
What tests can we add for this code? As it stands, looks like there are no tests, which means we won't know if we regress this?

RohitRanjanMS commented on src/WebJobs.Script/Config/ApplicationInsightsLoggerFilterOptionsSetup.cs:
```diff
@@ -0,0 +1,32 @@
+            // set this AsyncLocal to true to indicate the log should be skipped.
+            if (!_environment.IsDotNetInProc())
+            {
+                options.AddFilter<ApplicationInsightsLoggerProvider>((_, _) => !FilterApplicationInsightsFromWorker.Value);
```
we can remove one of the "_". Use the overload with just level filter.

brettsam replied:
Done.

brettsam commented:
@mathewc -- adding tests now. Required some shuffling around of how the GrpcCapabilities are created so I can inject this capability into an existing node test (even though node doesn't actually support it).

brettsam commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -112,7 +113,7 @@ internal class GrpcWorkerChannel : IRpcWorkerChannel, IDisposable
             _hostingConfigOptions = hostingConfigOptions;
 
-            _workerCapabilities = new GrpcCapabilities(_workerChannelLogger);
+            _grpcCapabilitiesFactory = grpcCapabilitiesFactory;
```
This change isn't great as some locations check capabilities before they're set, so now they're null... I'm going to think through this. The behavior of this capabilities object is pretty fragile (and I've just realized that we check/parse the values a lot... may have to refactor this first).

brettsam replied:
I ended up simplifying this behavior by adding a virtual method I can override to intercept these capabilities in a test. It's still not beautiful, but we have plans already to overhaul a bit of this (capabilities included) and I felt like I was getting too far into the weeds.

brettsam committed:
adding ability to ignore worker app insights logs via capabilities

brettsam commented:
My new PR backs some things out:
- @jviau's comment about testing filters was a great one. It highlighted some issues. I've moved this into an `ITelemetryProcessor`, which means now it definitely will only be touched by App Insights and will not be affected by other filters.
- Simplified the testing from other PRs. Rather than overhauling `GrpcCapabilities` here, I've filed an issue for some refactoring in the future: https://github.com/Azure/azure-functions-host/issues/8983.

kshyju commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannelFactory.cs:
```diff
@@ -25,12 +25,11 @@ public class GrpcWorkerChannelFactory : IRpcWorkerChannelFactory
 
         public GrpcWorkerChannelFactory(IScriptEventManager eventManager, IEnvironment environment, IRpcServer rpcServer, ILoggerFactory loggerFactory, IOptionsMonitor<LanguageWorkerOptions> languageWorkerOptions,
-            IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, IRpcWorkerProcessFactory rpcWorkerProcessManager, ISharedMemoryManager sharedMemoryManager, IFunctionDataCache functionDataCache,
+            IOptionsMonitor<ScriptApplicationHostOptions> applicationHostOptions, IRpcWorkerProcessFactory rpcWorkerProcessManager, ISharedMemoryManager sharedMemoryManager,
```
Should we also remove `languageWorkerOptions` and `rpcServer` params? They appears to be unused.

brettsam replied:
I didn't even notice that -- yeah, let me clean that up while I'm making these changes.

kshyju commented on src/WebJobs.Script/Environment/EnvironmentExtensions.cs:
```diff
@@ -606,5 +604,13 @@ public static bool IsTargetBasedScalingEnabled(this IEnvironment environment)
             return string.Equals(environment.GetEnvironmentVariable(TargetBaseScalingEnabled), "1");
         }
+
+        public static bool IsDotNetInProc(this IEnvironment environment)
```
Do we use this method? Don't see anywhere being called.

brettsam replied:
yeah, this likely doesn't really matter much anymore. will pull it out.

kshyju commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -402,9 +403,22 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
 
+            if (_environment.IsApplicationInsightsAgentEnabled() ||
+                (bool.TryParse(_workerCapabilities.GetCapabilityState(RpcWorkerConstants.WorkerApplicationInsightsLoggingEnabled), out bool appInsightsWorkerEnabled) &&
+                appInsightsWorkerEnabled))
```
For other places, we don't really check the capability value is truthy.(ex: line 398). Should we stick to that approach for being consistent?

brettsam replied:
I really dislike that we only check for existence of the key in a lot of places -- it means if you explicitly set something to "false", it'll still be evaluated as "true". I find that really confusing.

I didn't want to touch all that, though -- but I do think we should address it at some point. I just didn't want to add to the confusion.

kshyju commented on src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs:
```diff
@@ -402,9 +403,22 @@ internal void WorkerInitResponse(GrpcEvent initEvent)
                 ScriptHost.IsFunctionDataCacheEnabled = false;
             }
 
+            if (_environment.IsApplicationInsightsAgentEnabled() ||
```
Question about the `||` condition here. This means the app setting cannot override the worker behavior (_worker sent a capability indicating ai logging enabled_). But customer wants to override it for some reason without redeploying worker payload and this _or_ condition will not allow that. Is that intentional behavior?

brettsam replied:
Yeah, it was intentional. My thinking was that if either of these are set, we need to respect that and use it. In reality, most people won't even know about this and it's up to the workers/infrastructure to set these.

brettsam committed:
PR comments
